clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -741,12 +972,88 @@ impl<'a, 'b> TypeChecker<'a, 'b> {
+            AtomValue(ref value) | LiteralValue(ref value) => TypeSignature::type_of(value),
+            Atom(ref name) => self.lookup_variable(name, context)?,
+            List(ref expression) => self.type_check_function_application(expression, context)?,
+            TraitReference(_, _) | Field(_) => {
+                return Err(CheckErrors::UnexpectedTraitOrFieldReference.into());
+            }
+        };
+
+        match (expected_type, &expr.expr) {
+            (
+                TypeSignature::TraitReferenceType(expected_trait_id),
+                LiteralValue(Value::Principal(PrincipalData::Contract(ref contract_identifier))),
+            ) => {
+                // When a principal literal is used as a trait, make sure it implements the trait.
+                let contract_to_check = self
+                    .db
+                    .load_contract(&contract_identifier)
+                    .ok_or(CheckErrors::NoSuchContract(contract_identifier.to_string()))?;
+
+                let expected_trait =
+                    &lookup_trait(self.db, Some(&self.contract_context), expected_trait_id)?;
+                contract_to_check.check_trait_compliance(expected_trait_id, expected_trait)?;
+            }
+            (TypeSignature::TraitReferenceType(expected_trait_id), _) => {
+                // When any other expression is used as a trait, those with trait
+                // types should be checked for compatibility. Others should report
+                // an error.
+                let expr_trait_id =
+                    if let TypeSignature::TraitReferenceType(expr_trait_id) = expr_type {
+                        expr_trait_id
+                    } else {
+                        return Err(CheckErrors::TypeError(expected_type.clone(), expr_type).into());
+                    };
+                let actual_trait =
+                    lookup_trait(self.db, Some(&self.contract_context), &expr_trait_id)?;
+                let expected_trait =
+                    lookup_trait(self.db, Some(&self.contract_context), expected_trait_id)?;
+                trait_check_trait_compliance(
+                    self.db,
+                    Some(&self.contract_context),
+                    &expr_trait_id,
+                    &actual_trait,
+                    expected_trait_id,
+                    &expected_trait,
+                )?;
+            }
+            (_, _) => {
+                inner_type_check_type(
+                    self.db,
+                    Some(&self.contract_context),
+                    &expr_type,
+                    expected_type,
+                )?;
+            }
+        }
+
+        // If we reach here with no errors, then the expression can be
+        // treated as the expected type.
+        expr_type = expected_type.clone();
+
+        runtime_cost(
+            ClarityCostFunction::AnalysisTypeAnnotate,
+            self,
+            expr_type.type_size()?,
```

Comment:
I see we're doing lots of DB work above to load up traits.  Should that be assessed here as well?

Reply:
The costs haver all been added.